### ISSUE TITLE
@safe and derelict __gshared data

### ISSUE THREAD

**Issue by denizzzka** [DerelictOrg/DerelictPQ#17]
If I mark as @safe block what uses DerelictPQ functions all theese functions cause compile errors:
```
src/dpq2/result.d(155,16): Error: safe function 'OID' cannot access __gshared data 'PQftype'
src/dpq2/result.d(168,20): Error: safe function 'columnNum' cannot access __gshared data 'PQfnumber'
src/dpq2/result.d(180,25): Error: safe function 'columnName' cannot access __gshared data 'PQfname'
src/dpq2/result.d(189,37): Error: safe function 'dpq2.result.Answer.columnName' cannot call system function 'std.string.fromStringz'
src/dpq2/result.d(286,16): Error: safe function 'size' cannot access __gshared data 'PQgetlength'
src/dpq2/result.d(296,16): Error: safe function 'isNULL' cannot access __gshared data 'PQgetisnull'
src/dpq2/result.d(303,34): Error: safe function 'opIndex' cannot access __gshared data 'PQgetvalue'
```

**Comment by aldacron** [comment]
The function pointers in all of the Derelict bindings are declared as __gshared because they must be available across all threads, and __gshared variables cannot be accessed in @safe code. If there were declared as shared instead of __gshared, they would be accessible, but I don't know what implications that would have. Moreover, that would only apply to acting the function pointer variables themselves. Actually calling the functions through he pointers still requires the functions to be declared as @safe or @trusted before they can be called in @safe code. That's not going to happen, because it means I would have to go through the source of every C library Derelict binds to (not just DerelictPQ) and check to make sure they aren't doing anything that D considers unsafe. It's possible to declare C functions as @safe on the D side, but as a general rule they should always be the default @system unless they can be proven otherwise.

In short, none of the bindings in Derelict are ever going to be @safe or @trusted, so you can't use them in @safe code.

**Comment by denizzzka** [comment]
Thanks for explanation!

**Comment by nordlow** [comment]
> The function pointers in all of the Derelict bindings are declared as __gshared because they must be available across all threads. __gshared variables cannot be accessed in @safe code. If they were declared as shared instead of __gshared, they would be accessible, but I don't know what implications that would have. Moreover, that would only apply to accessing the function pointer variables themselves. Actually calling the functions through the pointers still requires the functions to be declared as @safe or @trusted before they can be called in @safe code. That's not going to happen, because it means I would have to go through the source of every C library Derelict binds to (not just DerelictPQ) and check to make sure they aren't doing anything that D considers unsafe. It's possible to declare C functions as @safe on the D side, but as a general rule they should always be the default @System unless they can be proven otherwise.
> 
> In short, none of the bindings in Derelict are ever going to be @safe or @trusted, so you can't use them in @safe code.

These function pointers are only mutated in the `static shared this` being by definition thread-safe (run only once per process).  After that they aren't modified. So calling these functions is really thread-safe. Afaict, these functions themselves should be qualified with shared.

**Comment by mdparker** [comment]
It's irrelevant at this point. I'm no longer maintaining Derelict. I also use `__gshared` in BindBC though, and there I'm not using a module ctor to load them. I don't see that changing to shared will buy me anything.

**Comment by nordlow** [comment]
> It's irrelevant at this point. I'm no longer maintaining Derelict. I also use `__gshared` in BindBC though, and there I'm not using a module ctor to load them. I don't see that changing to shared will buy me anything.

You can qualify them as `static shared` instead of `__gshared`.
